master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala

```diff
@@ -653,11 +678,39 @@ private[celeborn] class Master(
       override def run(): Unit = {
         statusSystem.handleAppLost(appId, requestId)
         logInfo(s"Removed application $appId")
+        if (hasHDFSStorage) {
+          cleanExpiredAppDirsOnHDFS(appId)
+        }
         context.reply(ApplicationLostResponse(StatusCode.SUCCESS))
       }
     })
   }
 
+  private def cleanExpiredAppDirsOnHDFS(dir: String = ""): Unit = {
+    if (hadoopFs == null) {
+      hadoopFs = CelebornHadoopUtils.getHadoopFS(conf)
+    }
+    val hdfsWorkPath = new Path(conf.hdfsDir, conf.workerWorkingDir)
+    if (hadoopFs.exists(hdfsWorkPath)) {
+      if (!dir.isEmpty) {
+        // delete specific app dir on application lost
+        hadoopFs.delete(new Path(hdfsWorkPath, dir), true)
+      } else {
+        val iter = hadoopFs.listStatusIterator(hdfsWorkPath)
+        while (iter.hasNext) {
+          val startTime = System.currentTimeMillis()
+          val fileStatus = iter.next()
+          if (!statusSystem.appHeartbeatTime.containsKey(fileStatus.getPath.getName)) {
+            hadoopFs.delete(fileStatus.getPath, true)
```

Comment:
we should try catch the `IOException` and print the failed to delete dir in case HDFS is abnormal or some permission deny issues

Reply:
Updated.